iceoryx_posh/source/roudi/port_manager.cpp

```diff
@@ -215,6 +219,158 @@ void PortManager::doDiscoveryForSubscriberPort(SubscriberPortType& subscriberPor
+            }
+        }
+        else
+        {
+            // protocol error
+            errorHandler(
+                Error::kPORT_MANAGER__HANDLE_CLIENT_PORTS_INVALID_CAPRO_MESSAGE, nullptr, iox::ErrorLevel::MODERATE);
+        }
+    });
+}
+
+void PortManager::makeAllServerPortsToStopOffer() noexcept
+{
+    for (auto port : m_portPool->getServerPortDataList())
+    {
+        port->m_offeringRequested.store(false, std::memory_order_relaxed);
+
+        popo::ServerPortRouDi serverPort(*port);
+        doDiscoveryForServerPort(serverPort);
+    }
+}
+
+void PortManager::destroyServerPort(popo::ServerPortRouDi::MemberType_t* const serverPortData) noexcept
+{
+    cxx::Ensures(serverPortData != nullptr && "serverPortData must not be a nullptr");
+
+    // create temporary server ports to orderly shut this server down
+    popo::ServerPortRouDi serverPortRoudi{*serverPortData};
+    popo::ServerPortUser serverPortUser{*serverPortData};
+
+    serverPortRoudi.releaseAllChunks();
+    serverPortUser.stopOffer();
+
+    // process STOP_OFFER for this server in RouDi and distribute it
+    serverPortRoudi.tryGetCaProMessage().and_then([this, &serverPortRoudi](auto caproMessage) {
+        cxx::Ensures(caproMessage.m_type == capro::CaproMessageType::STOP_OFFER);
+
+        m_portIntrospection.reportMessage(caproMessage);
+        /// @todo iox-#27 add server to service registry
+        // this->removeEntryFromServiceRegistry(caproMessage.m_serviceDescription);
+        this->sendToAllMatchingClientPorts(caproMessage, serverPortRoudi);
+        /// @todo iox-#27 forward server port to interface ports
+        // this->sendToAllMatchingInterfacePorts(caproMessage);
+    });
+
+    /// @todo iox-#27 report to port introspection
+
+    // delete server port from list after STOP_OFFER was processed
+    m_portPool->removeServerPort(serverPortData);
+
+    LogDebug() << "Destroyed server port";
+}
+
+void PortManager::handleServerPorts() noexcept
+{
+    // get the changes of server port offer state
+    for (auto serverPortData : m_portPool->getServerPortDataList())
+    {
+        popo::ServerPortRouDi serverPort(*serverPortData);
+
+        doDiscoveryForServerPort(serverPort);
+
+        // check if we have to destroy this server port
+        if (serverPort.toBeDestroyed())
+        {
+            destroyServerPort(serverPortData);
+        }
+    }
+}
+
+void PortManager::doDiscoveryForServerPort(popo::ServerPortRouDi& serverPort) noexcept
+{
+    serverPort.tryGetCaProMessage().and_then([this, &serverPort](auto caproMessage) {
+        /// @todo iox-#27 report to port instrospection
+
+        if (capro::CaproMessageType::OFFER == caproMessage.m_type)
+        {
+            /// @todo iox-#27 add to service registry?
+        }
+        else if (capro::CaproMessageType::STOP_OFFER == caproMessage.m_type)
+        {
+            /// @todo iox-#27 remove from service registry
+        }
+        else
+        {
+            // protocol error
```

Comment:
Maybe we should also add a `LogError` here with process name and service description.